weave-js/src/components/PagePanelComponents/Home/Browse3/feedback/HumanFeedback/HumanFeedback.tsx

```diff
@@ -0,0 +1,680 @@
+
+  const parsedRef = parseRef(columnRef);
+  const rawValues = Object.values(combinedFeedback)
+    .map(payload => {
+      const pRecord = payload as Record<
+        string,
+        Record<string, string | number>
+      >;
+      return pRecord[parsedRef.artifactName]?.[parsedRef.artifactVersion];
+    })
+    .filter(Boolean);
+
+  const viewerFeedbackVal =
+    combinedFeedback[viewer ?? '']?.[parsedRef.artifactName]?.[
+      parsedRef.artifactVersion
+    ];
+
+  // Get most recent value from the first feedback (since they're sorted by created_at desc)
+  const mostRecentVal =
+    foundFeedback[0]?.payload.value?.[parsedRef.artifactName]?.[
+      parsedRef.artifactVersion
+    ] ?? null;
+
+  return {
+    rawValues,
+    // Currently unused, but likely useful in the future
+    viewerFeedbackVal,
+    mostRecentVal,
+    combinedFeedback,
+  };
+};
+
+type FeedbackRequestProps = {
+  entity: string;
+  project: string;
+  viewer: string | null;
+  callRef: string;
+  feedbackSpecRef: string;
+  value: any;
+};
+
+// Utility function for creating feedback request
+const generateFeedbackRequestPayload = ({
+  entity,
+  project,
+  viewer,
+  callRef,
+  feedbackSpecRef,
+  value,
+}: FeedbackRequestProps) => {
+  const parsedRef = parseRef(feedbackSpecRef);
+  const humanAnnotationPayload: HumanAnnotationPayload = {
+    value: {
+      [parsedRef.artifactName]: {
+        [parsedRef?.artifactVersion]: value,
+      },
+    },
+  };
+  const feedbackType = makeAnnotationFeedbackType(parsedRef.artifactName);
+  const baseRequest = {
+    project_id: `${entity}/${project}`,
+    weave_ref: callRef,
+    creator: viewer,
+    feedback_type: feedbackType,
+    annotation_ref: feedbackSpecRef,
+    payload: humanAnnotationPayload,
+    sort_by: [{created_at: 'desc'}],
+  };
+  return baseRequest;
+};
+
+const inferTypeFromJsonSchema = (jsonSchema: Record<string, any>) => {
+  if (FEEDBACK_TYPE_OPTIONS.includes(jsonSchema.type)) {
+    return jsonSchema.type;
+  }
+  toast(`Unknown feedback type from spec: ${JSON.stringify(jsonSchema)}`, {
+    type: 'warning',
+  });
+  return null;
+};
+
+export const NumericalFeedbackColumn = ({
+  min,
+  max,
+  onAddFeedback,
+  defaultValue,
+  focused,
+  isInteger,
+}: {
+  min: number;
+  max: number;
+  onAddFeedback?: (value: number | null) => Promise<boolean>;
+  defaultValue: number | null;
+  focused?: boolean;
+  isInteger?: boolean;
+}) => {
+  const debouncedFn = useMemo(
+    () =>
+      _.debounce((val: number | null) => onAddFeedback?.(val), DEBOUNCE_VAL),
+    [onAddFeedback]
+  );
+  useEffect(() => {
+    return () => {
+      debouncedFn.cancel();
+    };
+  }, [debouncedFn]);
+
+  return (
+    <NumericalTextField
+      value={defaultValue}
+      onChange={debouncedFn}
+      min={min}
+      max={max}
+      isInteger={isInteger}
+      autoFocus={focused}
+    />
+  );
+};
+
+export const TextFeedbackColumn = ({
```

Comment:
future: a lot of this could merge with the zod editor

Reply:
yeah